src/node_port.py

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for handling mongos Kubernetes resources for a single mongos pod."""
+
+import logging
+from functools import cached_property
+from ops.charm import CharmBase
+from lightkube.models.meta_v1 import ObjectMeta
+from lightkube.core.client import Client
+from lightkube.core.exceptions import ApiError
+from lightkube.resources.core_v1 import Pod, Service
+from lightkube.models.core_v1 import ServicePort, ServiceSpec
+
+
+logger = logging.getLogger(__name__)
+
+# default logging from lightkube httpx requests is very noisy
+logging.getLogger("lightkube").disabled = True
+logging.getLogger("lightkube.core.client").disabled = True
+logging.getLogger("httpx").disabled = True
+logging.getLogger("httpcore").disabled = True
+
+
+class NodePortManager:
+    """Manager for handling mongos Kubernetes resources for a single mongos pod."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        pod_name: str,
+        namespace: str,
+    ):
+        self.charm = charm
+        self.pod_name = pod_name
+        self.app_name = "-".join(pod_name.split("-")[:-1])
+        self.namespace = namespace
+        self.nodeport_service_name = f"{self.app_name}-nodeport"
+
+    @cached_property
+    def client(self) -> Client:
+        """The Lightkube client."""
+        return Client(  # pyright: ignore[reportArgumentType]
+            field_manager=self.pod_name,
+            namespace=self.namespace,
+        )
+
+    # --- GETTERS ---
+
+    def get_pod(self, pod_name: str = "") -> Pod:
+        """Gets the Pod via the K8s API."""
+        # Allows us to get pods from other peer units
+        pod_name = pod_name or self.pod_name
+
+        return self.client.get(
+            res=Pod,
+            name=self.pod_name,
+        )
+
+    def build_node_port_services(self, port: str) -> Service:
+        """Builds a ClusterIP service for initial client connection."""
+        pod = self.get_pod(pod_name=self.pod_name)
+        if not pod.metadata:
+            raise Exception(f"Could not find metadata for {pod}")
+
+        unit_id = self.charm.unit.name.split("/")[1]
+        return Service(
+            metadata=ObjectMeta(
+                name=f"{self.charm.app.name}-{unit_id}-external",
+                namespace=self.namespace,
+                # owned by the StatefulSet
+                ownerReferences=pod.metadata.ownerReferences,
```

Comment:
**todo:** Sorry I missed this in my initial review. If you're interested in learning about K8s GC - https://kubernetes.io/docs/concepts/architecture/garbage-collection/

When we scale-down, say 5 --> 3 units, K8s will keep the Services for the deleted 2 units around, **unless** the Services' `owner` is also deleted, whereupon K8s will clean it up.
This snippet sets the Service owner as the App/StatefulSet, so services would only be cleaned up when the whole app goes down.

Instead, you want to remove the Pod-specific Service when the Pod/Unit is scaled-down.
Do this:

```python
        pod = self.get_pod(pod_name=self.pod_name)
        if not pod.metadata:
            raise Exception(f"Could not find metadata for {pod}")

        unit_id = self.charm.unit.name.split("/")[1]
        return Service(
            metadata=ObjectMeta(
                name=f"{self.charm.app.name}-{unit_id}-external",
                namespace=self.namespace,
                # owned by the Pod
                ownerReferences=[
                    OwnerReference(
                        apiVersion=pod.apiVersion,
                        kind=pod.kind,
                        name=self.pod_name,
                        uid=pod.metadata.uid,
                        blockOwnerDeletion=False,
                    )
                ],
...
```

Reply:
epic, thanks a ton for this! Added this in with an additional comment